### ISSUE TITLE
Centralized routing infrastructure

### ISSUE THREAD

**Issue by sjrct** [Halibot/halibot#103]
This was rambled on and ranted about in IRC and the meetings somewhat. The essential idea is centralize the location of routing so there is no need for things like the github module specifying its destinations within its own config. This also allows for a more transparent implementation of containers and a move away from the mostly artificial separation between church and—  ...modules and agents.

The syntax is to have a top-level field `routes` which contains an array of strings of the format `"{origin} -> {destination 1}, {destination 2}, ..."`. The advantage of using an array rather than a dictionary is that order is significant, only the first matching route is used. The origin can be a regular expression that is matched against module/agent name.

**Comment by richteer** [comment]
TL;DR of more IRC rambling: this might need to get tweaked around quite a bit for eventual container support.

I suspect there's going to be two flavors of containers, "passive" and "active" containers. Passive containers will be purely routing logic (much like the example of the route you posted above). Active containers however, will probably be a `HalObject`, and thus have an eventloop and the capability to have more complex routing logic. Question is, where should those containers be defined?

In the case of containers being `HalObjects`, should they be defined with the other agents/modules? What about what modules are "contained" in them? The purpose of the routes feature was to have a central lookup point for where a message will be sent based on RI. Active containers however sit in a weird spot of being both module-ish and route-ish.

**Comment by richteer** [comment]
Proposal: we define a top-level key called `routing`, which is an object. For now, we should only support the sub-key `table`, which contains the list as defined in the issue text. I think this will reduce the amount of overhaul needed when we get to implementing containers.

**Comment by sjrct** [comment]
I am not fond of the the `routing` object with sub fields, it feels more complicated than it needs to be, and if we want a separate key for container routes than we can just have another top-level field called `container-routes` or something like that.

Also, syntax proposed in IRC by richteer that I think is the most elegant:
```json
"routes": [
  { "a": "b"
  , "c": [ "d", "e" ]
  },
  { "g": "e" }
]
```
